### PR TITLE
chore: refactor parsers to avoid usage of parseurl package

### PR DESCRIPTION
**Pull request by david-luna**
### Summary

Add own implementation of `parseurl` package to remove its conditional dependency which, in some situations, may cause build issues to consumers of the agent.

Fixes: #1884

### Description

This PR refactors `parseUrl` function inside parsers module to accept URL strings and also request objects. The goal is to keep the login encapsulated into the parsers module and provide same output as the [parseurl package](https://www.npmjs.com/package/parseurl) conditionally imported into [express-utils file](https://github.com/elastic/apm-agent-nodejs/blob/34cc2a77820ae5b52f5ffab36f33fbaece8ccaf6/lib/instrumentation/express-utils.js#L13) 

This compatibility is ensured by adding the proper tests. There is one thing to highlight which is the different behavior between `parseurl` and node's `URL` object when it comes to authorisation like strings. As an example for the string `//todo@txt`
- `parseurl` considers it as the path part
- `URL` does not and path is empty

In this case I kept the `parseurl` package logic as requested in the issue but IMHO it can be removed. The path resolution logic is used when we want it for [transaction names](https://github.com/elastic/apm-agent-nodejs/blob/34cc2a77820ae5b52f5ffab36f33fbaece8ccaf6/lib/instrumentation/express-utils.js#L45) and in these scenarios I guess the requests contain in their url property more than a single auth like string.

### Checklist

- [x] Implement code
- [x] Add tests
- [x] Commit message follows [commit guidelines](https://github.com/elastic/apm-agent-nodejs/blob/main/CONTRIBUTING.md#commit-message-guidelines)

### CODE REVIEW

**Review by trentm — CHANGES_REQUESTED** [review]
Thanks for this PR! This went deeper than I expected. Sorry for the wall of text in response.

I went back through the history to try to find out why apm-agent-nodejs *has* this try/catch.
The decision to drop the `parseurl` dep, but conditionally depend on it was made here:
https://github.com/elastic/apm-agent-nodejs/pull/907#pullrequestreview-214493853
There were some Pros/Cons listed above in that same PR:
https://github.com/elastic/apm-agent-nodejs/pull/907#discussion_r264763308
If I read that correctly, it came down to a possible small performance benefit
(not having to re-parse `req.url`) if the user otherwise used `parseurl`.
We're changing that decision here. All good so far.

Justification for changing the decision:
1. The conditional import is a pain for bundlers.
2. The complexity of two subtley different URL parsing code paths isn't worth this possible small perf benefit. (From some experience with profiling this APM agent, the URL parsing has never shown up as a item of concern.)
3. (This is a new one.) We are testing one code path and many users are getting another. In testing of this APM agent we have `parseurl` installed -- as a side-effect of having 'express', 'koa', and some other packages installed for testing. A user of 'elastic-apm-node' will *not* have `parseurl` installed (express & koa depend on it; fastify, hapi, etc. do not).


# different behaviours

Thanks for showing the case where `new URL` and `parseurl()` differ -- the "TODO" in your patch.

```js
  // TODO: parseurl and URL have different APIs here
  // parseurl resolves it as a path
  // URL resolves it as absolute
  t.test('should not choke on auth-looking URL', function (t) {
    var req = createRequest('//todo@txt')
    var url = parsers.parseUrl(req)
    console.log(url)
    t.equals(url.pathname, '//todo@txt')
    t.end()
  })
```

This got me looking into this case more. `req.url` is defined here: https://nodejs.org/api/http.html#messageurl
This isn't actually a "URL". It is the `request-target` of the HTTP request line, as defined here: https://www.rfc-editor.org/rfc/rfc7230#section-3.1.1
All its possible forms are defined at https://www.rfc-editor.org/rfc/rfc7230#section-5.3

The Node.js docs section says:

> To parse the URL into its parts:
> ```
> new URL(request.url, `http://${request.headers.host}`);
> ```

Unfortunately, that advice is *wrong* for this case. For example, a request to `http://localhost:3000//todo@txt` *should* result in `pathname === '//todo@txt'`, but using the above results in:

```
> var request = { url: '//todo@txt', headers: { host: 'localhost:3000' }}
undefined
> new URL(request.url, `http://${request.headers.host}`);
URL {
  href: 'http://todo@txt/',
  origin: 'http://txt',
  protocol: 'http:',
  username: 'todo',
  password: '',
  host: 'txt',
  hostname: 'txt',
  port: '',
  pathname: '/',
  search: '',
  searchParams: URLSearchParams {},
  hash: ''
}
```

Basically the `parseurl` module is really about parsing an HTTP request line's `request-target`. Its implementation makes the assumption that if the `req.url` startswith '/', then it starts with the pathname: https://github.com/pillarjs/parseurl/blob/0376c2bc30920bc7435eb9413a587ea50875d000/index.js#L96-L100
It advertises that it yields the same result as `url.parse()`, but that isn't true in this case.


# what this PR could do

The only use case we have for `parseurl` is for `getPathFromRequest`. I.e. we only care about the `pathname` and don't need to fully replicate `parseurl`'s behaviour.  What about replacing the last part of `getPathFromRequest` with something like this?

```js
  if (usePathAsTransactionName) {
    // `req.url` is the "request-target" from the HTTP request line
    // (https://www.rfc-editor.org/rfc/rfc7230#section-3.1.1). The special-case
    // handling for leading '/' is because this:
    //    GET //foo/bar HTTP/1.1
    //    Host: example.com
    // should result in:
    //    { host: 'example.com', pathname: '//foo/bar' }
    // and *not*:
    //    { host: 'foo', pathname: '/bar' }
    //
    // Assuming 'http://' for the `base` URL is fine, because we don't use the
    // protocol.
    const base = 'http://' + (req.headers && req.headers.host)
    const parsed = (req.url.startsWith('/') ? new URL(base + req.url) : new URL(req.url, base))
    return parsed && parsed.pathname
  }
```

Then "parsers.js" could be left untouched. The `parseUrl` in "parsers.js" could probably stand to be refactored away, or at least documented. That it uses a constant `relative:///` and without a comment explaining why is weird. However, that is for a separate change.

It would be good to add a "test/instrumentation/express-utils.test.js" to test this new case.

Thoughts?

File: lib/instrumentation/express-utils.js
Comment (david-luna) [reply lib/instrumentation/express-utils.js:54]: At 1st I thought to have separated APIs for string and request types but since the return types was exactly the same I decided to overload the method (not sure if this is allowed in your coding guidelines).

Based on your comment about leveraging `headers.host` to get the base for the URL resolution I tihnk the line would be as you described above

```
const parsed = (req.url.startsWith('/') ? new URL(base + req.url) : new URL(req.url, base))
```

First think that came to my mind was to provide the base as a param of the `parseUrl` function with a default value of `relative:///` but since you mentioned there is no comment about why the constant is there I prefer to leave it untouched.

Maybe my lack of context about what a `Transaction` is makes me wonder if customers that will get new behavior of the parsing logic, hence different transaction names, are going to be impacted somehow.  I guess if transaction is an entity that lives within a session is fine but not if the entity has bigger scope than the session.

If what is mentioned above is not an issue I'd go with the proposed changes:
- leave parsers untouched & remove tests (there is no point to test URL node API)
- leverage host header & base constructor param in `getPathFromRequest`
- add tests for `getPathFromRequest` method

Thanks for the thorough review @trentm and I wait for you feedback and 👍 to apply your suggestions.

Comment (trentm) [reply lib/instrumentation/express-utils.js:54]: > Maybe my lack of context about what a `Transaction` is makes me wonder if customers that will get new behavior of the parsing logic, hence different transaction names

(If it helps, a "Transaction" is Elastic APM's name for a top-level tracing "span" in a service. In this case a Transaction is used to hold the tracing data for an incoming HTTP request. https://www.elastic.co/guide/en/apm/guide/current/data-model-transactions.html provides some details.)

I think adding tests for `getPathFromRequest` could cover all the same cases that "parseurl/test/test.js" covers for `parseurl(req)`. If there are any differences using `const parsed = (req.url.startsWith('/') ? new URL(base + req.url) : new URL(req.url, base))`, then we can discuss again. My guess is that all cases will give the same result with the newer implementation -- which means there will be no change in `transaction.name` behaviour for users.

Also, I think generally moving to the `URL` class for URL parsing and away from `url.parse`, `Url`, and the `parseurl` module is a good thing. https://nodejs.org/api/url.html#urlparseurlstring-parsequerystring-slashesdenotehost says

> url.parse() uses a lenient, non-standard algorithm for parsing URL strings. It is prone to security issues such as [host name spoofing](https://hackerone.com/reports/678487) and incorrect handling of usernames and passwords. Do not use with untrusted input. CVEs are not issued for url.parse() vulnerabilities. Use the [WHATWG URL](https://nodejs.org/api/url.html#the-whatwg-url-api) API instead.

I don't think the APM agent's usage of these is a concern for *security* issues, but the newer API is standardized and maintained.

> If what is mentioned above is not an issue I'd go with the proposed changes:
> ...

👍  Yes, it would be great if you could change your PR to do these.  One more thing is that once `parseurl` usage is removed you can apply this change for the `npm run test:deps` task:

```diff
diff --git a/package.json b/package.json
index 4ced2900..f96618b4 100644
--- a/package.json
+++ b/package.json
@@ -15,7 +15,7 @@
     "lint:yaml-files": "./dev-utils/lint-yaml-files.sh  # requires node >=10",
     "coverage": "COVERAGE=true ./test/script/run_tests.sh",
     "test": "./test/script/run_tests.sh",
-    "test:deps": "dependency-check index.js start.js start-next.js 'lib/**/*.js' 'test/**/*.js' '!test/activation-method/fixtures' '!test/instrumentation/modules/next/a-nextjs-app' --no-dev -i async_hooks -i perf_hooks -i parseurl -i node:http -i @azure/functions-core",
+    "test:deps": "dependency-check index.js start.js start-next.js 'lib/**/*.js' 'test/**/*.js' '!test/activation-method/fixtures' '!test/instrumentation/modules/next/a-nextjs-app' --no-dev -i async_hooks -i perf_hooks -i node:http -i @azure/functions-core",
     "test:tav": "tav --quiet && (cd test/instrumentation/modules/next/a-nextjs-app && tav --quiet)",
     "test:docs": "./test/script/docker/run_docs.sh",
     "test:types": "tsc --project test/types/tsconfig.json && tsc --project test/types/transpile/tsconfig.json && node test/types/transpile/index.js && tsc --project test/types/transpile-default/tsconfig.json && node test/types/transpile-default/index.js",
```

Comment (david-luna) [reply lib/instrumentation/express-utils.js]: Thanks @trentm for the clarification about transaction model :) 

Suggested changes are applied and pushed for you review

**Comment by trentm** [comment]
This looks great now.   I'll approve this and see if the "apm-ci/pr-merge" check can run now -- this is the set of tests that are run in our Jenkins-based CI.


For local testing I made this change:

```
diff --git a/examples/trace-http.js b/examples/trace-http.js
index f9fbb524..31e5840b 100755
--- a/examples/trace-http.js
+++ b/examples/trace-http.js
@@ -56,7 +56,12 @@ server.listen(3000, function () {
   //
   // Note that this there is no current "transaction" here, so this HTTP
   // request is not captured by APM. See "trace-http-request.js" for more.
-  const clientReq = http.request('http://localhost:3000/', function (clientRes) {
+  // const clientReq = http.request('http://localhost:3000/', function (clientRes) {
+  const clientReq = http.request({
+    host: 'localhost',
+    port: 3000,
+    path: '//user@foo'
+  }, function (clientRes) {
     console.log('client response: %s %s', clientRes.statusCode, clientRes.headers)
     const chunks = []
     clientRes.on('data', function (chunk) {
```

To test having an incoming HTTP request with the path case we are talking about. The generated "transaction" sent to APM server is:

```json
    {
        "transaction": {
            "name": "GET //user@foo",
            "type": "request",
            "result": "HTTP 2xx",
            "id": "80c4855cb1b44e8c",
            "trace_id": "e16442d9ff198b356f592c4dcda279df",
            "duration": 417.824,
            "timestamp": 1675277660776035,
            "sampled": true,
            "context": {
                "user": {},
                "tags": {},
                "custom": {},
                "service": {},
                "cloud": {},
                "message": {},
                "request": {
                    "http_version": "1.1",
                    "method": "GET",
                    "url": {
                        "raw": "//user@foo",
                        "protocol": "http:",
                        "hostname": "foo",
                        "port": "3000",
                        "full": "http://foo:3000"
                    },
                    "headers": {
                        "host": "localhost:3000",
                        "connection": "close"
                    },
                    "socket": {
                        "remote_address": "::ffff:127.0.0.1"
                    }
                },
                "response": {
                    "status_code": 200,
                    "headers": {
                        "server": "example-trace-http",
                        "content-type": "text/plain",
                        "content-length": "4",
                        "date": "Wed, 01 Feb 2023 18:54:21 GMT",
                        "connection": "close"
                    }
                }
            },
            "span_count": {
                "started": 1
            },
            "outcome": "success",
            "sample_rate": 1
        }
    }
```

I.e. the `transaction.name` is what we want.

### Aside

A separate issue is that the `transaction.context.request.url.*` is *wrong*:

```
                    "url": {
                        "raw": "//user@foo",
                        "protocol": "http:",
                        "hostname": "foo",
                        "port": "3000",
                        "full": "http://foo:3000"
                    },
```

However, that's a separate issue. That's coming from here: https://github.com/elastic/apm-agent-nodejs/blob/34cc2a77820ae5b52f5ffab36f33fbaece8ccaf6/lib/parsers.js#L27
I'll open a separate issue for that.

**Review by trentm — APPROVED** [review]

**Comment by trentm** [comment]
@elasticmachine, run elasticsearch-ci/docs

**Comment by trentm** [comment]
Thanks very much!

**Comment by david-luna** [comment]
Thanks @trentm for your feedback on this. Looking forward to collaborate again in another issue :)

**Pull request merged**
